<a href="https://colab.research.google.com/github/DavidCastro88/ProjectsData/blob/main/DataTransformCleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Data Cleaning Netflix dataset

***Questions:***

How many null values ​​do you find in the data? Can you delete them?

How many incomplete values ​​do you find in the data? Can you replace them?

Can you eliminate columns that do not provide you with information? Which are? Why would you delete them?

What type of data is the “release_year” column? Can you convert it to integer?

The “listed_in” column contains different values ​​separated by commas. Can you create a column and keep the first value?

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

In [ ]:
data = pd.read_csv('https://raw.githubusercontent.com/DavidCastro88/ProjectsData/main/Datasets/netflix1.csv')
data.head()

show_id     type                             title         director  \
0      s1    Movie              Dick Johnson Is Dead  Kirsten Johnson   
1      s3  TV Show                         Ganglands  Julien Leclercq   
2      s6  TV Show                     Midnight Mass    Mike Flanagan   
3     s14    Movie  Confessions of an Invisible Girl    Bruno Garotti   
4      s8    Movie                           Sankofa     Haile Gerima   

         country date_added  release_year rating  duration  \
0  United States  9/25/2021          2020  PG-13    90 min   
1         France  9/24/2021          2021  TV-MA  1 Season   
2  United States  9/24/2021          2021  TV-MA  1 Season   
3         Brazil  9/22/2021          2021  TV-PG    91 min   
4  United States  9/24/2021          1993  TV-MA   125 min   

                                           listed_in  
0                                      Documentaries  
1  Crime TV Shows, International TV Shows, TV Act...  
2                 TV Dramas, TV Horror, TV Mysteries  
3                 Children & Family Movies, Comedies  
4   Dramas, Independent Movies, International Movies

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8790 entries, 0 to 8789
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8790 non-null   object
 1   type          8790 non-null   object
 2   title         8790 non-null   object
 3   director      8790 non-null   object
 4   country       8790 non-null   object
 5   date_added    8790 non-null   object
 6   release_year  8790 non-null   int64 
 7   rating        8790 non-null   object
 8   duration      8790 non-null   object
 9   listed_in     8790 non-null   object
dtypes: int64(1), object(9)
memory usage: 686.8+ KB


### **1- How many null values ​​do you find in the data? Can you delete them?**

In [ ]:
data.isna().sum()

show_id         0
type            0
title           0
director        0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
dtype: int64

In [ ]:
data['type'].unique()

array(['Movie', 'TV Show'], dtype=object)

In [ ]:
data['director'].value_counts()

director
Not Given                         2588
Rajiv Chilaka                       20
Alastair Fothergill                 18
Raúl Campos, Jan Suter              18
Suhas Kadav                         16
                                  ... 
Matt D'Avella                        1
Parthiban                            1
Scott McAboy                         1
Raymie Muzquiz, Stu Livingston       1
Mozez Singh                          1
Name: count, Length: 4528, dtype: int64

In director column we have found 2588 registers without director, with "Not Given".

In [ ]:
data['country'].value_counts()

country
United States     3240
India             1057
United Kingdom     638
Pakistan           421
Not Given          287
                  ... 
Iran                 1
West Germany         1
Greece               1
Zimbabwe             1
Soviet Union         1
Name: count, Length: 86, dtype: int64

In the country column we have found 287 resgitres without country name, with "Not Given"

In [ ]:
data['date_added'].value_counts()

date_added
1/1/2020      110
11/1/2019      91
3/1/2018       75
12/31/2019     74
10/1/2018      71
             ... 
6/26/2015       1
6/23/2015       1
6/1/2015        1
5/29/2015       1
4/1/2014        1
Name: count, Length: 1713, dtype: int64

In [ ]:
dict(data['release_year'].value_counts())

{2018: 1146,
 2017: 1030,
 2019: 1030,
 2020: 953,
 2016: 901,
 2021: 592,
 2015: 555,
 2014: 352,
 2013: 286,
 2012: 236,
 2010: 192,
 2011: 185,
 2009: 152,
 2008: 135,
 2006: 96,
 2007: 88,
 2005: 80,
 2004: 64,
 2003: 59,
 2002: 51,
 2001: 45,
 1999: 39,
 1997: 38,
 2000: 37,
 1998: 36,
 1993: 28,
 1995: 25,
 1996: 24,
 1992: 23,
 1994: 22,
 1990: 22,
 1988: 18,
 1982: 17,
 1991: 17,
 1989: 16,
 1981: 13,
 1986: 13,
 1984: 12,
 1979: 11,
 1980: 11,
 1983: 11,
 1985: 10,
 1973: 10,
 1976: 9,
 1987: 8,
 1978: 7,
 1974: 7,
 1977: 7,
 1975: 7,
 1967: 5,
 1971: 5,
 1972: 5,
 1960: 4,
 1945: 4,
 1968: 3,
 1962: 3,
 1955: 3,
 1944: 3,
 1943: 3,
 1958: 3,
 1970: 2,
 1963: 2,
 1969: 2,
 1956: 2,
 1954: 2,
 1965: 2,
 1964: 2,
 1946: 2,
 1942: 2,
 1966: 1,
 1959: 1,
 1925: 1,
 1947: 1,
 1961: 1}

In [ ]:
data['rating'].value_counts()

rating
TV-MA       3205
TV-14       2157
TV-PG        861
R            799
PG-13        490
TV-Y7        333
TV-Y         306
PG           287
TV-G         220
NR            79
G             41
TV-Y7-FV       6
NC-17          3
UR             3
Name: count, dtype: int64

In [ ]:
dict(data['duration'].value_counts())

{'1 Season': 1791,
 '2 Seasons': 421,
 '3 Seasons': 198,
 '90 min': 152,
 '97 min': 146,
 '93 min': 146,
 '94 min': 146,
 '91 min': 144,
 '95 min': 137,
 '96 min': 130,
 '92 min': 129,
 '102 min': 122,
 '98 min': 120,
 '99 min': 118,
 '101 min': 116,
 '88 min': 116,
 '103 min': 114,
 '106 min': 111,
 '100 min': 108,
 '89 min': 106,
 '104 min': 104,
 '86 min': 103,
 '87 min': 101,
 '105 min': 101,
 '107 min': 98,
 '110 min': 97,
 '4 Seasons': 94,
 '108 min': 87,
 '116 min': 80,
 '112 min': 74,
 '85 min': 73,
 '109 min': 69,
 '113 min': 69,
 '111 min': 68,
 '84 min': 67,
 '118 min': 65,
 '83 min': 65,
 '5 Seasons': 64,
 '119 min': 63,
 '81 min': 62,
 '117 min': 61,
 '115 min': 60,
 '114 min': 56,
 '120 min': 56,
 '121 min': 54,
 '82 min': 52,
 '124 min': 52,
 '127 min': 48,
 '122 min': 45,
 '78 min': 45,
 '123 min': 44,
 '126 min': 44,
 '80 min': 43,
 '133 min': 42,
 '128 min': 41,
 '130 min': 40,
 '135 min': 39,
 '137 min': 38,
 '132 min': 37,
 '125 min': 36,
 '75 min': 35,
 '79 min': 3

In [ ]:
dict(data['listed_in'].value_counts())

As we can observe, only the columns 'country' and 'director' have null values with a total of 2588 and 287 respectively. However, it is not convenient to drop these values because important information for the later analysis would be lost.

### ***2. How many incomplete values ​​do you find in the data? Can you replace them?***

In [ ]:
data[(data['director']=='Not Given') | (data['country']=='Not Given')]

show_id     type                             title  \
12      s852    Movie                  99 Songs (Tamil)   
17        s4  TV Show             Jailbirds New Orleans   
18       s15  TV Show   Crime Stories: India Detectives   
23        s7    Movie  My Little Pony: A New Generation   
24       s12  TV Show                  Bangkok Breaking   
...      ...      ...                               ...   
8785   s8797  TV Show                        Yunus Emre   
8786   s8798  TV Show                         Zak Storm   
8787   s8801  TV Show                Zindagi Gulzar Hai   
8788   s8784  TV Show                              Yoko   
8789   s8786  TV Show                               YOM   

                           director        country  date_added  release_year  \
12                        Not Given       Pakistan   5/21/2021          2021   
17                        Not Given       Pakistan   9/24/2021          2021   
18                        Not Given       Pakistan   9/22/2021          2021   
23    Robert Cullen, José Luis Ucha      Not Given   9/24/2021          2021   
24                Kongkiat Komesiri      Not Given   9/23/2021          2021   
...                             ...            ...         ...           ...   
8785                      Not Given         Turkey   1/17/2017          2016   
8786                      Not Given  United States   9/13/2018          2016   
8787                      Not Given       Pakistan  12/15/2016          2012   
8788                      Not Given       Pakistan   6/23/2018          2016   
8789                      Not Given       Pakistan    6/7/2018          2016   

     rating   duration                                          listed_in  
12    TV-14    131 min     Dramas, International Movies, Music & Musicals  
17    TV-MA   1 Season                             Docuseries, Reality TV  
18    TV-MA   1 Season       British TV Shows, Crime TV Shows, Docuseries  
23       PG     91 min                           Children & Family Movies  
24    TV-MA   1 Season  Crime TV Shows, International TV Shows, TV Act...  
...     ...        ...                                                ...  
8785  TV-PG  2 Seasons                  International TV Shows, TV Dramas  
8786  TV-Y7  3 Seasons                                           Kids' TV  
8787  TV-PG   1 Season  International TV Shows, Romantic TV Shows, TV ...  
8788   TV-Y   1 Season                                           Kids' TV  
8789  TV-Y7   1 Season                                           Kids' TV  

[2875 rows x 10 columns]

In [ ]:
len(data[data['country'].isin(['United States','India','United Kingdom'])])/len(data)*100

56.14334470989761

We found a total of 2875 rows with incomplete values. It is possible to replace them by using APIs that have historical information and can retrieve the missing values, especially for the author. It is not recommended to use mode or sampling because the records are unique and heterogeneous.

### ***3- Can you eliminate columns that do not provide you with information? Which are? Why would you delete them?***

In general, we can eliminate any columns, depending of analysis purpose. The columns that can eliminate is show_id, because no is show relevant information. Also in our case, go to delete the columns country and release_year, beacause only we want characteristics directly relationship with netflix.

In [ ]:
data.drop(columns=['show_id','country','release_year'],axis=True,inplace=True)

In [ ]:
data

type                             title         director  date_added  \
0       Movie              Dick Johnson Is Dead  Kirsten Johnson   9/25/2021   
1     TV Show                         Ganglands  Julien Leclercq   9/24/2021   
2     TV Show                     Midnight Mass    Mike Flanagan   9/24/2021   
3       Movie  Confessions of an Invisible Girl    Bruno Garotti   9/22/2021   
4       Movie                           Sankofa     Haile Gerima   9/24/2021   
...       ...                               ...              ...         ...   
8785  TV Show                        Yunus Emre        Not Given   1/17/2017   
8786  TV Show                         Zak Storm        Not Given   9/13/2018   
8787  TV Show                Zindagi Gulzar Hai        Not Given  12/15/2016   
8788  TV Show                              Yoko        Not Given   6/23/2018   
8789  TV Show                               YOM        Not Given    6/7/2018   

     rating   duration                                          listed_in  
0     PG-13     90 min                                      Documentaries  
1     TV-MA   1 Season  Crime TV Shows, International TV Shows, TV Act...  
2     TV-MA   1 Season                 TV Dramas, TV Horror, TV Mysteries  
3     TV-PG     91 min                 Children & Family Movies, Comedies  
4     TV-MA    125 min   Dramas, Independent Movies, International Movies  
...     ...        ...                                                ...  
8785  TV-PG  2 Seasons                  International TV Shows, TV Dramas  
8786  TV-Y7  3 Seasons                                           Kids' TV  
8787  TV-PG   1 Season  International TV Shows, Romantic TV Shows, TV ...  
8788   TV-Y   1 Season                                           Kids' TV  
8789  TV-Y7   1 Season                                           Kids' TV  

[8790 rows x 7 columns]

### ***4- What type of data is the “release_year” column? Can you convert it to integer?***

The type of data, of "release_year" ia a integer, but we eliminated it.  

### ***5- The “listed_in” column contains different values ​​separated by commas. Can you create a column and keep the first value?***

In [ ]:
data['listed_in']

0                                           Documentaries
1       Crime TV Shows, International TV Shows, TV Act...
2                      TV Dramas, TV Horror, TV Mysteries
3                      Children & Family Movies, Comedies
4        Dramas, Independent Movies, International Movies
                              ...                        
8785                    International TV Shows, TV Dramas
8786                                             Kids' TV
8787    International TV Shows, Romantic TV Shows, TV ...
8788                                             Kids' TV
8789                                             Kids' TV
Name: listed_in, Length: 8790, dtype: object

In [ ]:
data['first_category'] = data['listed_in'].str.partition(',')[0]
data

type                             title         director  date_added  \
0       Movie              Dick Johnson Is Dead  Kirsten Johnson   9/25/2021   
1     TV Show                         Ganglands  Julien Leclercq   9/24/2021   
2     TV Show                     Midnight Mass    Mike Flanagan   9/24/2021   
3       Movie  Confessions of an Invisible Girl    Bruno Garotti   9/22/2021   
4       Movie                           Sankofa     Haile Gerima   9/24/2021   
...       ...                               ...              ...         ...   
8785  TV Show                        Yunus Emre        Not Given   1/17/2017   
8786  TV Show                         Zak Storm        Not Given   9/13/2018   
8787  TV Show                Zindagi Gulzar Hai        Not Given  12/15/2016   
8788  TV Show                              Yoko        Not Given   6/23/2018   
8789  TV Show                               YOM        Not Given    6/7/2018   

     rating   duration                                          listed_in  \
0     PG-13     90 min                                      Documentaries   
1     TV-MA   1 Season  Crime TV Shows, International TV Shows, TV Act...   
2     TV-MA   1 Season                 TV Dramas, TV Horror, TV Mysteries   
3     TV-PG     91 min                 Children & Family Movies, Comedies   
4     TV-MA    125 min   Dramas, Independent Movies, International Movies   
...     ...        ...                                                ...   
8785  TV-PG  2 Seasons                  International TV Shows, TV Dramas   
8786  TV-Y7  3 Seasons                                           Kids' TV   
8787  TV-PG   1 Season  International TV Shows, Romantic TV Shows, TV ...   
8788   TV-Y   1 Season                                           Kids' TV   
8789  TV-Y7   1 Season                                           Kids' TV   

                first_category  
0                Documentaries  
1               Crime TV Shows  
2                    TV Dramas  
3     Children & Family Movies  
4                       Dramas  
...                        ...  
8785    International TV Shows  
8786                  Kids' TV  
8787    International TV Shows  
8788                  Kids' TV  
8789                  Kids' TV  

[8790 rows x 8 columns]

In [ ]:
dict(data['first_category'].value_counts())

{'Dramas': 1599,
 'Comedies': 1210,
 'Action & Adventure': 859,
 'Documentaries': 829,
 'International TV Shows': 773,
 'Children & Family Movies': 605,
 'Crime TV Shows': 399,
 "Kids' TV": 385,
 'Stand-Up Comedy': 334,
 'Horror Movies': 275,
 'British TV Shows': 252,
 'Docuseries': 220,
 'Anime Series': 174,
 'International Movies': 128,
 'Reality TV': 120,
 'TV Comedies': 119,
 'Classic Movies': 80,
 'TV Dramas': 67,
 'Thrillers': 65,
 'Movies': 53,
 'TV Action & Adventure': 39,
 'Stand-Up Comedy & Talk Shows': 34,
 'Romantic TV Shows': 32,
 'Anime Features': 21,
 'Independent Movies': 20,
 'Classic & Cult TV': 20,
 'Music & Musicals': 18,
 'TV Shows': 16,
 'Sci-Fi & Fantasy': 13,
 'Cult Movies': 12,
 'TV Horror': 11,
 'Romantic Movies': 3,
 'Spanish-Language TV Shows': 2,
 'LGBTQ Movies': 1,
 'TV Sci-Fi & Fantasy': 1,
 'Sports Movies': 1}

In [70]:
data.drop(columns=['listed_in'],inplace=True)

In [71]:
data['date_added'] = pd.to_datetime(data['date_added'])

To continue, i´m going to pull apart Movie and TV Show, for later use in Power BI.

In [72]:
data_movie = data[data['type']=='Movie']
data_tvshow = data[data['type']=='TV Show']

In [73]:
data_movie

type                             title         director date_added  \
0     Movie              Dick Johnson Is Dead  Kirsten Johnson 2021-09-25   
3     Movie  Confessions of an Invisible Girl    Bruno Garotti 2021-09-22   
4     Movie                           Sankofa     Haile Gerima 2021-09-24   
6     Movie                      The Starling   Theodore Melfi 2021-09-24   
7     Movie   Motu Patlu in the Game of Zones      Suhas Kadav 2021-05-01   
...     ...                               ...              ...        ...   
8702  Movie                          The Bund        Not Given 2018-09-20   
8707  Movie                  The Darkest Dawn        Not Given 2018-06-23   
8716  Movie                  The Great Battle        Not Given 2019-04-08   
8763  Movie       Twisted Trunk, Big Fat Body        Not Given 2017-01-15   
8783  Movie              Yoko and His Friends        Not Given 2018-06-23   

     rating duration            first_category  
0     PG-13   90 min             Documentaries  
3     TV-PG   91 min  Children & Family Movies  
4     TV-MA  125 min                    Dramas  
6     PG-13  104 min                  Comedies  
7     TV-Y7   87 min  Children & Family Movies  
...     ...      ...                       ...  
8702  TV-14  103 min        Action & Adventure  
8707  TV-MA   75 min        Action & Adventure  
8716  TV-MA  136 min        Action & Adventure  
8763  TV-14   89 min                    Dramas  
8783   TV-Y   78 min  Children & Family Movies  

[6126 rows x 7 columns]

In [74]:
data_movie.loc[:, 'duration'] = data_movie['duration'].str.partition(' ')[0]

In [75]:
data_movie.rename({'duration':'duration(min)'},inplace=True,axis=1)

<ipython-input-75-8281f7ecf7d0>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_movie.rename({'duration':'duration(min)'},inplace=True,axis=1)


In [76]:
data_movie

type                             title         director date_added  \
0     Movie              Dick Johnson Is Dead  Kirsten Johnson 2021-09-25   
3     Movie  Confessions of an Invisible Girl    Bruno Garotti 2021-09-22   
4     Movie                           Sankofa     Haile Gerima 2021-09-24   
6     Movie                      The Starling   Theodore Melfi 2021-09-24   
7     Movie   Motu Patlu in the Game of Zones      Suhas Kadav 2021-05-01   
...     ...                               ...              ...        ...   
8702  Movie                          The Bund        Not Given 2018-09-20   
8707  Movie                  The Darkest Dawn        Not Given 2018-06-23   
8716  Movie                  The Great Battle        Not Given 2019-04-08   
8763  Movie       Twisted Trunk, Big Fat Body        Not Given 2017-01-15   
8783  Movie              Yoko and His Friends        Not Given 2018-06-23   

     rating duration(min)            first_category  
0     PG-13            90             Documentaries  
3     TV-PG            91  Children & Family Movies  
4     TV-MA           125                    Dramas  
6     PG-13           104                  Comedies  
7     TV-Y7            87  Children & Family Movies  
...     ...           ...                       ...  
8702  TV-14           103        Action & Adventure  
8707  TV-MA            75        Action & Adventure  
8716  TV-MA           136        Action & Adventure  
8763  TV-14            89                    Dramas  
8783   TV-Y            78  Children & Family Movies  

[6126 rows x 7 columns]

In [78]:
data_tvshow.loc[:, 'duration'] = data_tvshow['duration'].str.partition(' ')[0]

In [80]:
data_tvshow.rename({'duration':'duration(seasons)'},inplace=True,axis=1)

<ipython-input-80-f1d9f8008da3>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_tvshow.rename({'duration':'duration(seasons)'},inplace=True,axis=1)


In [81]:
data_tvshow

type                            title         director date_added  \
1     TV Show                        Ganglands  Julien Leclercq 2021-09-24   
2     TV Show                    Midnight Mass    Mike Flanagan 2021-09-24   
5     TV Show    The Great British Baking Show  Andy Devonshire 2021-09-24   
17    TV Show            Jailbirds New Orleans        Not Given 2021-09-24   
18    TV Show  Crime Stories: India Detectives        Not Given 2021-09-22   
...       ...                              ...              ...        ...   
8785  TV Show                       Yunus Emre        Not Given 2017-01-17   
8786  TV Show                        Zak Storm        Not Given 2018-09-13   
8787  TV Show               Zindagi Gulzar Hai        Not Given 2016-12-15   
8788  TV Show                             Yoko        Not Given 2018-06-23   
8789  TV Show                              YOM        Not Given 2018-06-07   

     rating duration(seasons)          first_category  
1     TV-MA                 1          Crime TV Shows  
2     TV-MA                 1               TV Dramas  
5     TV-14                 9        British TV Shows  
17    TV-MA                 1              Docuseries  
18    TV-MA                 1        British TV Shows  
...     ...               ...                     ...  
8785  TV-PG                 2  International TV Shows  
8786  TV-Y7                 3                Kids' TV  
8787  TV-PG                 1  International TV Shows  
8788   TV-Y                 1                Kids' TV  
8789  TV-Y7                 1                Kids' TV  

[2664 rows x 7 columns]

We can observed that have a total of 2664 series, and 6126 movies.

In [82]:
data_tvshow.to_csv('data_series.csv', index=False)
data_movie.to_csv('data_movies.csv', index=False)